src/dns/mod.rs

```diff
@@ -26,6 +26,26 @@ pub trait ConfigureResolver {
     fn configure_resolver(&self, &mut ResolverOpts);
 }
 
+pub trait Refine {
+    type Future: Future<Item = RefinedName, Error = ResolveError>;
+    /// Attempts to refine `name` to a fully-qualified name.
+    ///
+    /// This method does DNS resolution for `name` and ignores the IP address
+    /// result, instead returning the `Name` that was resolved.
+    ///
+    /// For example, a name like `web` may be refined to `web.example.com.`,
+    /// depending on the DNS search path.
+    fn refine(&self, name: &Name) -> Self::Future;
+}
+
+pub trait MakeRefine: Send + 'static {
+    type Refine: Refine + Clone + Send + Sync + 'static;
+    fn make(self, resolver: &Resolver) -> Self::Refine;
```

Comment:
It's odd that we need to pass a resolver in here? I would expect this to encapsulate creation of the resolver?

Reply:
This is a hack and I don't care for it either. It's a workaround for the issue that the current DNS resolver's `new` returns both a resolver _and_ a background task. We can't just immediately spawn the background task in this function because we want to run it on the control thread, but we need the resolver before we start running the control thread.

Perhaps the right thing would be to do a deeper refactor of how this stuff works, but that requires further thought.